packages/zipkin/src/batch-recorder.js

```diff
@@ -0,0 +1,118 @@
+const {now} = require('./time');
+const thriftTypes = require('./gen-nodejs/zipkinCore_types');
+const {
+  MutableSpan,
+  Endpoint,
+  ZipkinAnnotation,
+  BinaryAnnotation
+} = require('./internalRepresentations');
+
+class BatchRecorder {
+  constructor({
+    logger,
+    timeout = 60 * 1000000 // default timeout = 60 seconds
+  }) {
+    this.logger = logger;
+    this.timeout = timeout;
+    this.partialSpans = new Map();
+
+    // read through the partials spans regularly
+    // and collect any timed-out ones
+    setInterval(() => {
+      this.partialSpans.forEach((span, id) => {
+        if (this._timedOut(span)) {
+          this._writeSpan(id);
+        }
+      });
+    }, 1000);
+  }
+
+  _writeSpan(id) {
+    const spanToWrite = this.partialSpans.get(id);
+    // ready for garbage collection
+    this.partialSpans.delete(id);
+    this.logger.logSpan(spanToWrite);
+  }
+
+  _updateSpanMap(id, updater) {
```

Comment:
Yes, it was actually ported (more or less) directly from Finagle to JavaScript. That's why there's even an option/some/none inplementation in there. We should make zipkin-js more JavaScript-idiomatic later on.